modules/client/src/main/java/org/apache/ignite/internal/client/tx/ClientTransactions.java

```diff
@@ -73,19 +78,35 @@ static CompletableFuture<ClientTransaction> beginAsync(
                 w -> {
                     w.out().packBoolean(readOnly);
                     w.out().packLong(timeout);
-                    w.out().packLong(observableTimestamp);
+                    w.out().packLong(observableTimestamp.get().longValue());
+                    if (!readOnly) {
+                        w.out().packInt(pm == null ? -1 : pm.tableId());
+                        w.out().packInt(pm == null ? -1 : pm.partition());
+                    }
                 },
-                r -> readTx(r, readOnly),
-                preferredNodeName,
+                r -> readTx(r, readOnly, pm, observableTimestamp, timeout),
+                pm == null ? null : pm.nodeConsistentId(),
+                null,
                 null,
                 false);
     }
 
-    private static ClientTransaction readTx(PayloadInputChannel r, boolean isReadOnly) {
+    private static ClientTransaction readTx(
+            PayloadInputChannel r,
+            boolean isReadOnly,
+            @Nullable PartitionMapping pm,
+            HybridTimestampTracker tracker,
+            long timeout
+    ) {
         ClientMessageUnpacker in = r.in();
 
         long id = in.unpackLong();
-
-        return new ClientTransaction(r.clientChannel(), id, isReadOnly);
+        if (isReadOnly) { // TODO we can split to client read/write txns.
+            return new ClientTransaction(r.clientChannel(), id, isReadOnly, EMPTY, null, EMPTY, null, timeout);
+        } else {
+            UUID txId = in.unpackUuid();
```

Comment:
Missing feature check - this will fail on older servers.

Reply:
🆗